File: scrape_kenpom.R
#!/usr/bin/env Rscript
# ==============================================================================
# KenPom Daily Scraper
# Pulls Adjusted Efficiency Margin (AdjEM), Adjusted Offense (AdjO),
# and Adjusted Defense (AdjD) for all D1 teams.
#
# Requires a KenPom subscription.
# Uses chromote (headless Chrome) to bypass Cloudflare bot protection.
#
# Dependencies:  install.packages(c("chromote", "rvest"))
#
# Credentials (in priority order):
#   1. Environment variables  KENPOM_EMAIL / KENPOM_PASSWORD
#      (set these as GitHub Secrets for the Actions workflow)
#   2. Local file  .kenpom_creds  with two lines:
#        email=you@example.com
#        password=yourpassword
#
# Usage:  Rscript scrape_kenpom.R
# Output: kenpom_ratings_YYYY-MM-DD.csv in the kenpom_data/ subfolder
# ==============================================================================

library(chromote)
library(rvest)

# --- Load credentials --------------------------------------------------------
# Prefer env vars (GitHub Actions), fall back to local creds file

email    <- Sys.getenv("KENPOM_EMAIL",    unset = "")
password <- Sys.getenv("KENPOM_PASSWORD", unset = "")

if (email == "" || password == "") {
  # Find script directory: works via Rscript, source(), or interactive RStudio
  script_dir <- tryCatch(dirname(sys.frame(1)$ofile), error = function(e) ".")
  creds_file <- file.path(script_dir, ".kenpom_creds")
  if (file.exists(creds_file)) {
    creds_raw <- readLines(creds_file, warn = FALSE)
    for (line in creds_raw) {
      parts <- strsplit(trimws(line), "=", fixed = TRUE)[[1]]
      if (length(parts) >= 2) {
        key <- parts[1]
        val <- paste(parts[-1], collapse = "=")  # handle = in password
        if (key == "email")    email    <- val
        if (key == "password") password <- val
      }
    }
  }
}

if (email == "" || password == "") {
  stop("KenPom credentials not found.\n",
       "Set KENPOM_EMAIL and KENPOM_PASSWORD env vars (or GitHub Secrets),\n",
       "or create a .kenpom_creds file with email=... and password=... lines.")
}

# --- script_dir fallback (needed for output later) --------------------------
if (!exists("script_dir")) {
  script_dir <- tryCatch(dirname(sys.frame(1)$ofile), error = function(e) ".")
}

# --- Login to KenPom via headless Chrome ------------------------------------

cat("Launching headless Chrome ...\n")

b <- ChromoteSession$new()

# Navigate to KenPom homepage
cat("  Navigating to kenpom.com ...\n")
b$Page$navigate("https://kenpom.com/")
b$Page$loadEventFired(timeout = 30)
Sys.sleep(2)  # let Cloudflare challenge resolve

# Fill in login form and submit
cat("  Logging in ...\n")
b$Runtime$evaluate(sprintf(
  'document.querySelector("#email").value = %s;
   document.querySelector("#password").value = %s;',
  jsonlite::toJSON(email, auto_unbox = TRUE),
  jsonlite::toJSON(password, auto_unbox = TRUE)
))

# Click the submit button
b$Runtime$evaluate('document.querySelector("input[type=submit], button[type=submit]").click();')

# Wait for navigation after login
Sys.sleep(3)
b$Page$loadEventFired(timeout = 30)
Sys.sleep(1)

# --- Fetch the ratings page --------------------------------------------------

cat("Fetching ratings table ...\n")

# Navigate to the main ratings page (may already be there after login redirect)
b$Page$navigate("https://kenpom.com/")
b$Page$loadEventFired(timeout = 30)
Sys.sleep(2)

# Get the full page HTML
result <- b$Runtime$evaluate("document.documentElement.outerHTML")
page_source <- result$result$value

# Close the browser session
b$close()

page_html <- read_html(page_source)

# --- Parse the ratings table --------------------------------------------------

# The main table has id "ratings-table"
ratings_node <- html_element(page_html, "#ratings-table")

if (is.na(ratings_node)) {
  # Fallback: try first big table on the page
  ratings_node <- html_element(page_html, "table")
}

if (is.na(ratings_node)) {
  stop("Could not find ratings table. Login may have failed or page structure changed.\n",
       "Check that your credentials are correct.")
}

raw_table <- html_table(ratings_node, header = TRUE)

# --- Clean up columns ---------------------------------------------------------

# KenPom table columns vary slightly but the core layout is:
# Rk | Team | Conf | W-L | AdjEM | AdjO | AdjO Rk | AdjD | AdjD Rk | ...
# Some columns share header names; we rename by position

# Standardize column names
n_cols <- ncol(raw_table)
cat(sprintf("  Parsed table with %d rows, %d columns\n", nrow(raw_table), n_cols))

# Find the columns we care about by looking at headers
headers <- names(raw_table)

# Clean the table: remove separator rows (KenPom inserts blank rows between ranks)
raw_table <- raw_table[!is.na(raw_table[[1]]) & raw_table[[1]] != "", ]
raw_table <- raw_table[!grepl("^\\s*$", raw_table[[1]]), ]

# Build clean output
# Column positions (typical kenpom layout):
#  1=Rk, 2=Team, 3=Conf, 4=W-L, 5=AdjEM, 6=AdjO, 7=AdjO_Rk, 8=AdjD, 9=AdjD_Rk, ...
ratings <- data.frame(
  rank   = as.integer(raw_table[[1]]),
  team   = trimws(raw_table[[2]]),
  conf   = trimws(raw_table[[3]]),
  record = trimws(raw_table[[4]]),
  adj_em = as.numeric(raw_table[[5]]),
  adj_o  = as.numeric(raw_table[[6]]),
  adj_d  = as.numeric(raw_table[[8]]),
  stringsAsFactors = FALSE
)

# Drop any rows that didn't parse (header repeats, separators, etc.)
ratings <- ratings[!is.na(ratings$rank), ]

# Remove seed numbers that KenPom appends to team names during tournament
ratings$team <- gsub("\\s*\\d+$", "", ratings$team)

cat(sprintf("  Cleaned to %d teams\n", nrow(ratings)))

# --- Save to CSV --------------------------------------------------------------

out_dir <- file.path(script_dir, "kenpom_data")
if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)

today    <- Sys.Date()
out_file <- file.path(out_dir, sprintf("kenpom_ratings_%s.csv", today))

write.csv(ratings, out_file, row.names = FALSE)
cat(sprintf("\nSaved %d teams to %s\n", nrow(ratings), out_file))

# --- Quick preview ------------------------------------------------------------

cat("\nTop 20 teams:\n")
cat(sprintf("  %3s  %-25s %-6s %6s %6s %6s\n",
            "Rk", "Team", "Conf", "AdjEM", "AdjO", "AdjD"))
cat(paste(rep("-", 62), collapse = ""), "\n")
for (i in 1:min(20, nrow(ratings))) {
  r <- ratings[i, ]
  cat(sprintf("  %3d  %-25s %-6s %6.1f %6.1f %6.1f\n",
              r$rank, r$team, r$conf, r$adj_em, r$adj_o, r$adj_d))
}

cat("\nDone.\n")

File: scrape_kenpom.py
#!/usr/bin/env python3
"""
KenPom Daily Scraper
Pulls AdjEM, AdjO, AdjD for all D1 teams using nodriver to bypass Cloudflare.

Credentials (in priority order):
  1. Environment variables KENPOM_EMAIL / KENPOM_PASSWORD
  2. Local file .kenpom_creds with email=... and password=... lines

Usage:  python3 scrape_kenpom.py
Output: kenpom_data/kenpom_ratings_YYYY-MM-DD.csv
"""

import asyncio
import csv
import os
import re
import sys
from datetime import date
from pathlib import Path

import nodriver as uc
from bs4 import BeautifulSoup


def load_credentials():
    email = os.environ.get("KENPOM_EMAIL", "")
    password = os.environ.get("KENPOM_PASSWORD", "")

    if not email or not password:
        script_dir = Path(__file__).parent
        creds_file = script_dir / ".kenpom_creds"
        if creds_file.exists():
            for line in creds_file.read_text().strip().splitlines():
                line = line.strip()
                if "=" in line:
                    key, val = line.split("=", 1)
                    if key == "email":
                        email = val
                    elif key == "password":
                        password = val

    if not email or not password:
        print("ERROR: KenPom credentials not found.", file=sys.stderr)
        print("Set KENPOM_EMAIL/KENPOM_PASSWORD env vars or create .kenpom_creds", file=sys.stderr)
        sys.exit(1)

    return email, password


def parse_ratings(html: str) -> list[dict]:
    soup = BeautifulSoup(html, "html.parser")
    table = soup.find("table", id="ratings-table")
    if not table:
        table = soup.find("table")
    if not table:
        print("ERROR: Could not find ratings table.", file=sys.stderr)
        sys.exit(1)

    tbody = table.find("tbody")
    rows = tbody.find_all("tr") if tbody else table.find_all("tr")

    ratings = []
    for row in rows:
        cells = row.find_all("td")
        if len(cells) < 9:
            continue

        rank_text = cells[0].get_text(strip=True)
        if not rank_text or not rank_text.isdigit():
            continue

        team_name = cells[1].get_text(strip=True)
        # Remove seed numbers appended during tournament
        team_name = re.sub(r"\s*\d+$", "", team_name)

        try:
            ratings.append({
                "rank": int(rank_text),
                "team": team_name,
                "conf": cells[2].get_text(strip=True),
                "record": cells[3].get_text(strip=True),
                "adj_em": float(cells[4].get_text(strip=True)),
                "adj_o": float(cells[5].get_text(strip=True)),
                "adj_d": float(cells[7].get_text(strip=True)),
            })
        except (ValueError, IndexError):
            continue

    return ratings


async def main():
    email, password = load_credentials()
    script_dir = Path(__file__).parent

    print("Launching browser...")
    browser = await uc.start(headless=True)

    print("Navigating to kenpom.com...")
    page = await browser.get("https://kenpom.com/")

    # Wait for Cloudflare to resolve
    print("Waiting for Cloudflare challenge...")
    for i in range(30):
        await asyncio.sleep(1)
        try:
            title = await page.evaluate("document.title")
            if "Just a moment" not in str(title):
                print(f"  Cloudflare resolved after {i+1}s")
                break
        except Exception:
            pass
    else:
        print("WARNING: Cloudflare may not have resolved after 30s, continuing anyway...")

    await asyncio.sleep(2)

    # Check if we need to login (ratings visible without login on kenpom)
    html = await page.evaluate("document.documentElement.outerHTML")

    if "ratings-table" not in html:
        print("No ratings table found, attempting login...")
        # Try to find and fill login form
        try:
            email_input = await page.find("email", best_match=True)
            await email_input.send_keys(email)
            pw_input = await page.find("password", best_match=True)
            await pw_input.send_keys(password)
            submit = await page.find("submit", best_match=True)
            await submit.click()
            await asyncio.sleep(3)
            html = await page.evaluate("document.documentElement.outerHTML")
        except Exception as e:
            print(f"Login attempt failed: {e}")

    if "ratings-table" not in html:
        print("ERROR: Still no ratings table after login attempt.", file=sys.stderr)
        # Save debug HTML
        debug_file = script_dir / "debug_page.html"
        debug_file.write_text(html)
        print(f"Debug HTML saved to {debug_file}", file=sys.stderr)
        browser.stop()
        sys.exit(1)

    print("Found ratings table, parsing...")
    ratings = parse_ratings(html)
    print(f"  Parsed {len(ratings)} teams")

    browser.stop()

    # Save to CSV
    out_dir = script_dir / "kenpom_data"
    out_dir.mkdir(exist_ok=True)
    today = date.today().isoformat()
    out_file = out_dir / f"kenpom_ratings_{today}.csv"

    with open(out_file, "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=["rank", "team", "conf", "record", "adj_em", "adj_o", "adj_d"])
        writer.writeheader()
        writer.writerows(ratings)

    print(f"\nSaved {len(ratings)} teams to {out_file}")

    # Quick preview
    print(f"\n{'Rk':>3}  {'Team':<25} {'Conf':<6} {'AdjEM':>6} {'AdjO':>6} {'AdjD':>6}")
    print("-" * 62)
    for r in ratings[:20]:
        print(f"{r['rank']:>3}  {r['team']:<25} {r['conf']:<6} {r['adj_em']:>6.1f} {r['adj_o']:>6.1f} {r['adj_d']:>6.1f}")

    print("\nDone.")


if __name__ == "__main__":
    uc.loop().run_until_complete(main())
